core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java

```diff
@@ -0,0 +1,148 @@
+package ca.jrvs.apps.grep;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.regex.Pattern;
+import org.apache.log4j.BasicConfigurator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JavaGrepImp implements JavaGrep {
+
+  final Logger logger = LoggerFactory.getLogger(JavaGrepImp.class);
+
+  private String regex;
+  private String rootPath;
+  private String outFile;
+
+  public static void main(String[] args) {
+    if (args.length != 3) {
+      throw new IllegalArgumentException("USAGE: JavaGrep regex rootPath outFile");
+    }
+    BasicConfigurator.configure();
+    JavaGrepImp javaGrepImp = new JavaGrepImp();
+    javaGrepImp.setRegex(args[0]);
+    javaGrepImp.setRootPath(args[1]);
+    javaGrepImp.setOutFile(args[2]);
+
+    try {
+      javaGrepImp.process();
+    } catch (Exception ex) {
+      javaGrepImp.logger.error(ex.getMessage(), ex);
+    }
+  }
+
+  @Override
+  public void process() throws IOException {
+    List<String> matchedLines = new LinkedList<>();
+    for (File file : listFiles(getRootPath())) {
+      for (String line : readLines(file)) {
+        if (containsPattern(line)) {
+          matchedLines.add(line);
+        }
+      }
+    }
+    writeToFile(matchedLines);
+  }
+
+  @Override
+  public List<File> listFiles(String rootDir) {
+    LinkedList<File> files = new LinkedList<>();
+    File file = new File(rootDir);
+    if (file.exists()) {
+      if (file.isFile()) {
+        files.add(file);
+        return files;
+      } else if (file.isDirectory()) {
+        File[] contents = file.listFiles();
+        if (contents != null) {
+          for (File f : contents) {
+            List<File> moreFiles = listFiles(f.getAbsolutePath());
+            if (moreFiles != null) {
+              files.addAll(moreFiles);
+            }
+          }
+          return files;
+        }
+      }
+    }
+    return null;
```

Comment:
try not to return null since the caller has to handle it (due to null pointer exception).

You can just return empty `files` at the end of the method.